adserver/tasks.py

```diff
@@ -143,3 +148,48 @@ def daily_update_impressions(day=None):
             AdImpression.objects.filter(pk=impression.pk).update(
                 **{impression_type: values["total"]}
             )
+
+
+@app.task()
+def daily_update_keywords(day=None):
+    """
+    Update the KeywordImpression index each day.
+
+    :arg day: An optional datetime object representing a day
+    """
+    start_date, end_date = _get_day(day)
+
+    log.info("Updating KeywordImpression for %s-%s", start_date, end_date)
+
+    for impression_type in IMPRESSION_TYPES:
+        queryset = _default_filters(impression_type, start_date, end_date)
+
+        for values in (
+            queryset.values("publisher", "advertisement", "keywords")
+            .annotate(total=Count("keywords"))
+            .filter(total__gt=0)
+            .order_by("-keywords")
+        ):
+
+            keywords = json.loads(values["keywords"])
+            ad = Advertisement.objects.filter(id=values["advertisement"]).first()
```

Comment:
Since you use the flight later, you could save a query (in the loop) with a `select_related` here.

Reply:
Thinking more on it, I think we might be able to return this data from the query. I'll see if that works reasonably well.